dd-sdk-android-gradle-plugin/src/main/kotlin/com/datadog/gradle/plugin/DdFileUploadTask.kt

```diff
@@ -0,0 +1,326 @@
+     * The url of the remote repository where the source code was deployed.
+     */
+    @get:Input
+    var remoteRepositoryUrl: String = ""
+
+    /**
+     * Build ID which will be used for mapping file matching.
+     */
+    @get:Input
+    var buildId: Provider<String> = providerFactory.provider { "" }
+
+    /**
+     * datadog-ci.json file, if found or applicable for the particular task.
+     */
+    @Optional
+    @get:InputFile
+    var datadogCiFile: File? = null
+
+    /**
+     * The sourceSet root folders.
+     */
+    @get:InputFiles
+    var sourceSetRoots: List<File> = emptyList()
+
+    /**
+     * The file containing the repository description.
+     */
+    @get:OutputFile
+    var repositoryFile: File = File("")
+
+    init {
+        group = DdAndroidGradlePlugin.DATADOG_TASK_GROUP
+        // it is never up-to-date, because request may fail
+        outputs.upToDateWhen { false }
+    }
+
+    @Internal
+    internal abstract fun getFilesList(): List<Uploader.UploadFileInfo>
+
+    /**
+     * Uploads the files retrieved from `getFilesList` to Datadog.
+     */
+    @TaskAction
+    fun applyTask() {
+        datadogCiFile?.let {
+            applyDatadogCiConfig(it)
+        }
+        applySiteFromEnvironment()
+        validateConfiguration()
+
+        check(!(apiKey.contains("\"") || apiKey.contains("'"))) {
+            INVALID_API_KEY_FORMAT_ERROR
+        }
+
+        check(buildId.isPresent && buildId.get().isNotEmpty()) {
+            MISSING_BUILD_ID_ERROR
+        }
+
+        val mappingFiles = getFilesList()
+        if (mappingFiles.isEmpty()) return
+
+        var repositories = emptyList<RepositoryInfo>()
+        repositoryDetector?.let {
+            repositories = it.detectRepositories(
+                sourceSetRoots,
+                remoteRepositoryUrl
+            )
+        }
+        if (repositories.isNotEmpty()) {
+            generateRepositoryFile(repositories)
+        }
+
+        val site = DatadogSite.valueOf(site)
+        var lastError: Exception? = null
+        for (mappingFile in mappingFiles) {
+            LOGGER.info("Uploading ${mappingFile.fileType} file: ${mappingFile.file.absolutePath}")
+            try {
+                uploader.upload(
+                    site,
+                    mappingFile,
+                    if (repositories.isEmpty()) null else repositoryFile,
+                    apiKey,
+                    DdAppIdentifier(
+                        serviceName = serviceName,
+                        version = versionName,
+                        versionCode = versionCode,
+                        variant = variantName,
+                        buildId = buildId.get()
+                    ),
+                    repositories.firstOrNull(),
+                    !disableGzipOption.isPresent
+                )
+            } catch (e: Exception) {
+                LOGGER.error("Failed to upload ${mappingFile.fileType} file", e)
+                lastError = e
```

Comment:
maybe instead we can collect all of them, and if there is only one, re-throw it as-is, and if there are multiple, create a new exception and add all of them to it using https://docs.oracle.com/javase/7/docs/api/java/lang/Throwable.html#addSuppressed%28java.lang.Throwable%29? wdyt? this way we are not losing any.

Reply:
Yeah that sounds good. I'll make that change.